File: frontend/src/services/statistics.ts
import api from './api';

// 类型定义
export interface CourseOverview {
  course_id: number;
  course_name: string;
  student_count: number;
  document_count: number;
  question_set_count: number;
}

export interface QuestionSetStatistics {
  question_set_id: number;
  title: string;
  total_assigned: number;
  completed_count: number;
  completion_rate: number;
  average_score: number | null;
  failed_count: number;
  failed_reasons: string[];
}

export interface StudentInfo {
  id: number;
  username: string;
  email: string;
  full_name: string;
  joined_at: string;
  is_active: boolean;
}

export interface StudentStatistics {
  total_courses: number;
  total_question_sets: number;
  completed_count: number;
  average_score: number | null;
}

// 趋势统计类型
export interface DailySubmission {
  date: string;
  count: number;
}

export interface SubmissionTrend {
  data: DailySubmission[];
  total: number;
}

export interface QuestionSetCompletion {
  id: number;
  title: string;
  total_assigned: number;
  completed_count: number;
  completion_rate: number;
  average_score: number | null;
}

export interface QuestionSetCompletionList {
  items: QuestionSetCompletion[];
}

export interface DailyScore {
  date: string;
  score: number | null;
  count: number;
}

export interface ScoreTrend {
  data: DailyScore[];
}

// 统计 API
export const statisticsApi = {
  // 课程概览
  courseOverview: (courseId: number) =>
    api.get<CourseOverview>(`/statistics/course/${courseId}/overview`),

  // 试题集统计
  questionSetStats: (questionSetId: number) =>
    api.get<QuestionSetStatistics>(`/statistics/question-set/${questionSetId}`),

  // 课程学生列表
  courseStudents: (courseId: number) =>
    api.get<StudentInfo[]>(`/statistics/course/${courseId}/students`),

  // 学生个人统计
  myStatistics: () =>
    api.get<StudentStatistics>('/statistics/my-statistics'),

  // 答题提交趋势
  submissionTrend: (courseId: number, days: number = 7) =>
    api.get<SubmissionTrend>(`/statistics/course/${courseId}/submission-trend`, {
      params: { days },
    }),

  // 试题集完成情况
  questionSetCompletion: (courseId: number) =>
    api.get<QuestionSetCompletionList>(`/statistics/course/${courseId}/question-set-completion`),

  // 平均分趋势
  scoreTrend: (courseId: number, days: number = 7) =>
    api.get<ScoreTrend>(`/statistics/course/${courseId}/score-trend`, {
      params: { days },
    }),
};

File: backend/app/schemas/statistics.py
"""统计相关的请求/响应模型"""

from datetime import datetime

from pydantic import BaseModel


class CourseOverview(BaseModel):
    """课程概览"""

    course_id: int
    course_name: str
    student_count: int
    document_count: int
    question_set_count: int


class QuestionSetStatistics(BaseModel):
    """试题集统计"""

    question_set_id: int
    title: str
    total_assigned: int
    completed_count: int
    completion_rate: float
    average_score: float | None
    failed_count: int
    failed_reasons: list[str]


class StudentInfo(BaseModel):
    """学生信息"""

    id: int
    username: str
    full_name: str
    email: str
    joined_at: datetime
    is_active: bool


class StudentListResponse(BaseModel):
    """学生列表响应"""

    students: list[StudentInfo]
    total: int


class StudentStatistics(BaseModel):
    """学生个人统计"""

    total_courses: int
    total_question_sets: int
    completed_count: int
    average_score: float | None


class DailySubmission(BaseModel):
    """每日提交统计"""

    date: str
    count: int


class SubmissionTrend(BaseModel):
    """答题提交趋势"""

    data: list[DailySubmission]
    total: int


class QuestionSetCompletion(BaseModel):
    """试题集完成情况"""

    id: int
    title: str
    total_assigned: int
    completed_count: int
    completion_rate: float
    average_score: float | None


class QuestionSetCompletionList(BaseModel):
    """试题集完成情况列表"""

    items: list[QuestionSetCompletion]


class DailyScore(BaseModel):
    """每日平均分"""

    date: str
    score: float | None
    count: int


class ScoreTrend(BaseModel):
    """平均分趋势"""

    data: list[DailyScore]
